.github/workflows/test.yml

```diff
@@ -446,10 +457,25 @@ jobs:
             gcloud compute instances delete "${INSTANCE}" --zone "${{ env.ZONE }}" --delete-disks all --quiet
           fi
 
-      - name: Get disk state name from gcloud
-        id: get-disk-name
+      # If a triggering file for a rebuild was changed on a PR or we manually triggered a rebuild,
+      # use the latest available disk state as the regenerate-stateful-disks job was executed
+      - name: Get latest state disk name from gcloud
+        id: get-latest-disk-name
+        if: ${{ (needs.build.outputs.any_changed == 'true' && github.event_name == 'pull_request') || github.event.inputs.regenerate-disks == 'true' }}
+        run: |
+          GCP_STATE_DISK=$(gcloud compute images list --filter="name~zebrad-cache-${{ env.GITHUB_REF_SLUG_URL }} AND name~-checkpoint" --format="value(NAME)" --sort-by=~creationTimestamp --limit=1)
```

Comment:
Optional Performance: Do we want to re-use any matching state version here, rather than strictly matching the branch name?

If we match the branch name, we will rebuild the state twice for each database format PR:
* in the original PR
* in the Mergify queue PR

Reply:
Maybe this comments applies to the `regenerate-stateful-disks` job, from here on https://github.com/ZcashFoundation/zebra/blob/feat-state-versioning/.github/workflows/test.yml#L294

On a PR, so far, we've been re-building the state based on this condition:
`${{ steps.changed-files-specific.outputs.any_changed == 'true' || github.event.inputs.regenerate-disks == 'true'}}`

If the `constants.rs` is changed in a commit, the rebuild has been happening on each new commit on that specific PR and also in the Mergify queue.

> Do we want to re-use any matching state version here, rather than strictly matching the branch name?
@teor2345 I'm don't think I have enough knowledge on the behavior from using matching-version states, available from other PRs or if there's any edge-case that might cause syncing from genesis or an unexpected failure. I took this approach to ensure there's not state contamination between PR, but maybe it's not needed. I'll let you make that call and define the desired design (maybe in a ticket to get it done in another PR?)

Reply:
> If the `constants.rs` is changed in a commit, the rebuild has been happening on each new commit on that specific PR and also in the Mergify queue.

If we merge this PR, the performance of state rebuilds will be worse than the current CI behaviour. We will rebuild on every commit. But currently, we rebuild the cache once, then use it as long as the state version stays the same.

Checking the state version is a low priority, so I'm not sure if we should merge this PR. What do you think?

I would be happy to merge version checks that have similar performance.

> > Do we want to re-use any matching state version here, rather than strictly matching the branch name?
> 
> @teor2345 I don't think I have enough knowledge on the behavior from using matching-version states available from other PRs, or if there's any edge-case that might cause syncing from genesis, or an unexpected failure. I took this approach to ensure there's no state contamination between PRs, but maybe it's not needed.

The goal was to share cached state between PRs and commits, not to isolate the state from each commit. We need to do this for performance reasons - otherwise every `constants.rs` commit takes 3 hours.

I'm sorry for the confusion, hopefully we can be clearer next time. 😢 

> I'll let you make that call and define the desired design (maybe in a ticket to get it done in another PR?)

It seems like we have a few open low-priority DevOps PRs, and some of them will (temporarily) make CI performance worse.

But Zebra needs a cached state for the integration tests, and that is a high priority.

So I'm not sure what the best way forward is here.

I wonder if there is a quick fix that unblocks the high priority tasks, and lets us keep working on the low priority tasks when we have time? Or a quick change that avoids the performance issues?

Maybe we can catch up and talk through the details? Is the testing meeting tomorrow ok?

Reply:
Yeah, let's have a time tomorrow in the testing meeting to go through this one.

Reply:
Ok, well it seems like we should just merge this, and fix it up later?

I opened ticket #4153 for that fix, we can decide if anything depends on it at the testing meeting.

Reply:
(I'll approve the PR, if you're happy with this solution, feel free to resolve this comment and queue it for merge.)